File: src/interpreter/parser.ts
import { Token } from "./tokenizer";

interface Text {
  kind: "Text";
  speaker: string;
  text: string;
}
interface DisplayBackground {
  kind: "DisplayBackground";
  path: string;
}
export type Command = Text | DisplayBackground;

const Text = (speaker: string, text: string): Text => ({
  kind: "Text",
  speaker,
  text,
});
const DisplayBackground = (path: string): DisplayBackground => ({
  kind: "DisplayBackground",
  path,
});

export const generateCommands = (tokens: Token[]): Command[] => {
  let commands: Command[] = [];
  let queue = tokens;

  while (queue.length > 0) {
    if (
      queue[0]?.kind === "Text" &&
      queue[1]?.kind === "Text" &&
      queue[2]?.kind === "NewLine"
    ) {
      commands.push(Text(queue[0].text, queue[1].text));
      queue = queue.slice(3);
    } else if (queue[0]?.kind === "Text" && queue[1]?.kind === "NewLine") {
      commands.push(Text("", queue[0].text));
      queue = queue.slice(2);
    } else if (queue[0]?.kind === "Symbol" && queue[1]?.kind === "Symbol") {
      if (queue[0].text === "background") {
        commands.push(DisplayBackground(queue[1].text));
      } else {
        throw "Syntax error";
      }
      queue = queue.slice(2);
    } else if (queue[0]?.kind === "NewLine") {
      queue = queue.slice(1);
    } else {
      throw "Syntax error";
    }
  }

  return commands;
};

File: src/interpreter/parser.rs
use std::collections::{HashMap, VecDeque};

use crate::domain::GameData;

use super::tokenizer::{Token, TokenContainer};

#[derive(Clone)]
pub enum Command {
    Text {
        speaker: Option<String>,
        text: String,
    },
    CharacterText {
        speaker: String,
        text: String,
    },
    DisplayBackground {
        image_path: String,
    },
}

pub fn generate_commands(
    tokens: &Vec<TokenContainer>,
    game_data: &mut GameData,
) -> VecDeque<Command> {
    let mut queue: &[TokenContainer] = tokens;
    let mut commands = VecDeque::new();

    while queue.len() > 0 {
        match &queue[..] {
            [TokenContainer {
                token: Token::Text(speaker),
                ..
            }, TokenContainer {
                token: Token::Text(text),
                ..
            }, TokenContainer {
                token: Token::NewLine,
                ..
            }, ..] => {
                commands.push_back(Command::Text {
                    speaker: Some(speaker.clone()),
                    text: text.clone(),
                });
                queue = &queue[3..];
            }

            [TokenContainer {
                token: Token::Symbol(speaker),
                ..
            }, TokenContainer {
                token: Token::Text(text),
                ..
            }, TokenContainer {
                token: Token::NewLine,
                ..
            }, ..] => {
                commands.push_back(Command::CharacterText {
                    speaker: speaker.clone(),
                    text: text.clone(),
                });
                queue = &queue[3..];
            }

            [TokenContainer {
                token: Token::Text(text),
                ..
            }, TokenContainer {
                token: Token::NewLine,
                ..
            }, ..] => {
                commands.push_back(Command::Text {
                    speaker: None,
                    text: text.to_owned(),
                });
                queue = &queue[2..];
            }

            [TokenContainer {
                token: Token::Symbol(command),
                ..
            }, TokenContainer {
                token: Token::Symbol(target),
                ..
            }, ..] => {
                if command == "background" {
                    commands.push_back(Command::DisplayBackground {
                        image_path: target.clone(),
                    })
                } else {
                    panic!("Syntax error");
                }
                queue = &queue[2..];
            }

            [TokenContainer {
                token: Token::NewLine,
                ..
            }, ..] => queue = &queue[1..],

            [TokenContainer {
                token: Token::Symbol(name),
                ..
            }, TokenContainer {
                token: Token::Equals,
                ..
            }, TokenContainer {
                token: Token::ObjectStart,
                ..
            }, ..] => {
                let (new_queue, object) = parse_object(&queue[3..]);
                game_data.insert(name.clone(), object);
                queue = new_queue;
            }

            [..] => {
                if let Some(TokenContainer { location, .. }) = queue.first() {
                    panic!(
                        "Syntax error at location: (row: {}, col: {})",
                        location.row, location.col
                    );
                } else {
                    panic!("Syntax error");
                }
            }
        }
    }

    commands
}

fn parse_object(mut queue: &[TokenContainer]) -> (&[TokenContainer], HashMap<String, String>) {
    let mut map = HashMap::new();

    loop {
        if queue.len() == 0 {
            panic!("Finished parsing while creating object");
        }

        match &queue[..] {
            [TokenContainer {
                token: Token::ObjectEnd,
                ..
            }, ..] => {
                return (&queue[1..], map);
            }
            [TokenContainer {
                token: Token::Key(key),
                ..
            }, TokenContainer {
                token: Token::Symbol(val),
                ..
            }, ..] => {
                map.insert(key.clone(), val.clone());
                queue = &queue[2..];
            }
            [TokenContainer {
                token: Token::Key(key),
                ..
            }, TokenContainer {
                token: Token::Text(val),
                ..
            }, ..] => {
                map.insert(key.clone(), val.clone());
                queue = &queue[2..];
            }
            [TokenContainer {
                token: Token::NewLine,
                ..
            }, ..] => {
                queue = &queue[1..];
            }
            [TokenContainer {
                token: Token::Comma,
                ..
            }, ..] => {
                queue = &queue[1..];
            }
            _ => {
                if let Some(TokenContainer { location, .. }) = queue.first() {
                    panic!(
                        "Object creation error at location: (row: {}, col: {})",
                        location.row, location.col
                    );
                } else {
                    panic!("Object creation error");
                }
            }
        }
    }
}
